demos/ota/ota_demo_core_mqtt/ota_demo_core_mqtt.c

```diff
@@ -0,0 +1,1974 @@
+                     0,
+                     NULL,
+                     pdMS_TO_TICKS( MQTT_AGENT_MS_TO_WAIT_FOR_NOTIFICATION ) );
+
+    if( xCommandContext.xReturnStatus != MQTTSuccess )
+    {
+        LogError( ( "Failed to send unsubsribe packet to broker with error = %u.", xCommandContext.xReturnStatus ) );
+        xOtaMqttStatus = OtaMqttSubscribeFailed;
+    }
+    else
+    {
+        LogInfo( ( "Unsubsribed from topic %.*s to broker.\n\n",
+                   usTopicFilterLength,
+                   pcTopicFilter ) );
+    }
+
+    return xOtaMqttStatus;
+}
+/*-----------------------------------------------------------*/
+
+static void prvSetOtaInterfaces( OtaInterfaces_t * pxOtaInterfaces )
+{
+    /* Initialize OTA library OS Interface. */
+    pxOtaInterfaces->os.event.init = OtaInitEvent_FreeRTOS;
+    pxOtaInterfaces->os.event.send = OtaSendEvent_FreeRTOS;
+    pxOtaInterfaces->os.event.recv = OtaReceiveEvent_FreeRTOS;
+    pxOtaInterfaces->os.event.deinit = OtaDeinitEvent_FreeRTOS;
+    pxOtaInterfaces->os.timer.start = OtaStartTimer_FreeRTOS;
+    pxOtaInterfaces->os.timer.stop = OtaStopTimer_FreeRTOS;
+    pxOtaInterfaces->os.timer.delete = OtaDeleteTimer_FreeRTOS;
+    pxOtaInterfaces->os.mem.malloc = Malloc_FreeRTOS;
+    pxOtaInterfaces->os.mem.free = Free_FreeRTOS;
+
+    /* Initialize the OTA library MQTT Interface.*/
+    pxOtaInterfaces->mqtt.subscribe = prvMqttSubscribe;
+    pxOtaInterfaces->mqtt.publish = prvMqttPublish;
+    pxOtaInterfaces->mqtt.unsubscribe = prvMqttUnSubscribe;
+
+    /* Initialize the OTA library PAL Interface.*/
+    pxOtaInterfaces->pal.getPlatformImageState = otaPal_GetPlatformImageState;
+    pxOtaInterfaces->pal.setPlatformImageState = otaPal_SetPlatformImageState;
+    pxOtaInterfaces->pal.writeBlock = otaPal_WriteBlock;
+    pxOtaInterfaces->pal.activate = otaPal_ActivateNewImage;
+    pxOtaInterfaces->pal.closeFile = otaPal_CloseFile;
+    pxOtaInterfaces->pal.reset = otaPal_ResetDevice;
+    pxOtaInterfaces->pal.abort = otaPal_Abort;
+    pxOtaInterfaces->pal.createFile = otaPal_CreateFileForRx;
+}
+/*-----------------------------------------------------------*/
+
+static void prvOTAAgentTask( void * pParam )
+{
+    /* Calling OTA agent task. */
+    OTA_EventProcessingTask( pParam );
+    LogInfo( ( "OTA Agent stopped." ) );
+
+    vTaskDelete( NULL );
+}
+/*-----------------------------------------------------------*/
+
+static void prvMQTTAgentTask( void * pParam )
+{
+    BaseType_t xResult = pdFAIL;
+    MQTTStatus_t xMQTTStatus = MQTTSuccess;
+
+    ( void ) pParam;
+
+    do
+    {
+        /* MQTTAgent_CommandLoop() is effectively the agent implementation.  It
+         * will manage the MQTT protocol until such time that an error occurs,
+         * which could be a disconnect.  If an error occurs the MQTT context on
+         * which the error happened is returned so there can be an attempt to
+         * clean up and reconnect however the application writer prefers. */
```

Comment:
This isn't accurate any more
```suggestion
         * which could be a disconnect.  If an error occurs the MQTT error code
         * is returned and the queue left uncleared so there can be an attempt to
         * clean up and reconnect however the application writer prefers. */
```